### DISCUSSION TITLE
proxy / decorate a router

### DISCUSSION THREAD

**Discussion by kaminskypavel** [trpc/trpc#5572] [Q&A] [answered]
Hey everyone.

I'm building a chrome extension that has two adapters .. one is for communicating with the api server (lets call it an `apiRouter`) and another to communicate between the different parts(worlds) of the chrome extension (specifically the background script and a content script).

since `apiRouter `is only available for the background script , but the `extensionRouter ` is available for all the other parts.

I'd like the `extensionRouter ` to proxy the `apiRouter `, in such a way that a query()  from the api AppRouter would be available natively on the `extensionRouter `

my naive implementation was create a new proxy method for the client

```ts
export function createProxy<T>(api: any): T {
    return new Proxy(api, {
        get(target, propKey, receiver) {
            if (propKey === 'useQuery') {
                return function (...args) {
                    console.log(`Redirecting method useQuery to query with args:`, args);
                    return target.query(...args);
                };
            } else if (typeof target[propKey] === 'function') {
                return function (...args) {
                    console.log(`Calling ${propKey.toString()} with args:`, args);
                    return Reflect.get(target, propKey, receiver).apply(target, args);
                };
            } else if (typeof target[propKey] === 'object' && target[propKey] !== null) {
                console.log(`Creating proxy for ${propKey.toString()}`);
                return createProxy(target[propKey]);
            } else {
                console.log(`Returning property ${propKey.toString()}`);
                return target[propKey];
            }
        }
    });
}
```

but it looks like the proxy is being swapped by another implementation by tRPC.

what am I missing , is there a better way to tackle this?

**Comment by KATT** [comment]
I'm not sure I get the problem but you might want to look at [`createProxy.ts`](https://github.com/trpc/trpc/blob/495eca31ebfd5dc54eea42ecc05cd9e47e9cdb15/packages/server/src/unstable-core-do-not-import/createProxy.ts) and you can also use the [`TRPCUntypedClient`](https://github.com/trpc/trpc/blob/495eca31ebfd5dc54eea42ecc05cd9e47e9cdb15/packages/client/src/createTRPCUntypedClient.ts) in some cases, might be useful for rerouting

**Comment by kaminskypavel** [comment]
thanks @KATT , let me try to clear what im trying to achieve.

![image](https://github.com/trpc/trpc/assets/4253088/0b79afb3-9da2-4593-9987-6da5433cf20c)

Since final goal is to access `apiRouter` from `backgroundClient`, but the limitation it cannot access it directly.
see how i need to manually propagate every `query() / mutation()` through the `backgroundRouter`.

I was wondering what would be the elegant way to proxy `apiRouter` so the DX would feel more like this :

![image](https://github.com/trpc/trpc/assets/4253088/4e2f42fe-3278-4bf8-bdda-7bff53280aea)

**Reply by KATT** [reply, marked as answer]
I see what you're getting at. I think that's totally doable but I don't have the bandwidth to do it for you

I think you can do something like:

- Create custom handler in the extension that can receive trpc procedure calls that will act as a gateway proxy to the inner api client
  - (this handler could prob be a _procedure_ inside the "outer" api)
- This custom handler has it's own trpc client (probably using `TRPCUntypedClient` and pass messages on)
- Once you get the above working, you need to export the types of the inner API
- When you have the inner api, you can stitch this together using the same principle as the [SOA-approach](https://github.com/trpc/trpc/tree/next/examples/soa)
  - splitLink that:
    - knows how to talk to your existing trpc api 
    - if it's a call that should be relayed into the "inner" trpc api, you need a custom links that can communicate to the gateway

**Reply by kaminskypavel** [reply]
just wanted to say a huge thank you.

following the [soa-example](https://github.dev/trpc/trpc/blob/next/examples/soa/client/client.ts) I was able to come up with an even simpler solution. 

(this probably can be achieved with splitLink as well)

```ts

// * heavily based on the discussion here
// * https://github.com/trpc/trpc/discussions/5572#discussioncomment-8844371
// * https://github.dev/trpc/trpc/blob/next/examples/soa/client/client.ts


const combiedLink = (runtime: TRPCClientRuntime) => {
    // initialize the different links for different targets
    const services = {
        api: apiLink(runtime),
        background: backgroundLink(runtime),
    };
    return (ctx: {op: Operation<unknown>; next: (op: Operation<unknown>) => OperationResultObservable<AnyRouter, unknown>;}) => {
        const {op} = ctx;
        // split the path by `.` as the first part will signify the server target name
        const pathParts = op.path.split('.');

        // first part of the query should be `server1` or `server2`
        const serviceName = pathParts.shift() as keyof typeof services;

        // combine the rest of the parts of the paths
        // -- this is what we're actually calling the target server with
        const path = pathParts.join('.');
        console.log(`> calling ${serviceName} on path ${path}`, {
            input: op.input,
        });

        const link = services[serviceName];
        console.log({link, path, op});
        return link({
            ...ctx,
            op: {
                ...op,
                // override the target path with the prefix removed
                path,
            },
        });
    };
};
```


this make the DX so much better, thanks again for your direction @KATT 🙏
